### PR TITLE
Module compounds

### PR DESCRIPTION
**Pull request by hpcdgrie**
Create module compounds via python script or let the mapeditor write the script by selecting the modules you want to combine-> right click-> save as module group
Scripts in .config/vistle and the .vsl ending will automatically loaded when starting Vistle. Use this mechanism to permanently add your compounds to the module browser.

### CODE REVIEW

File: lib/vistle/control/hub.cpp [hunk 1]
```diff
@@ -2307,7 +2419,23 @@ bool Hub::handlePriv(const message::ExecutionDone &done)
     if (m_isMaster) {
         auto notif(done);
         notif.setDestId(Id::Broadcast);
-        sendAll(notif);
+        auto parent = getParentCompound(done.senderId());
+        if (parent != message::Id::Invalid) {
+            const auto &av = m_stateTracker.getStaticModuleInfo(parent);
+            bool isExecuing = false;
```
Comment (aumuell) [inline lib/vistle/control/hub.cpp:2425]: isExecuting

File: lib/vistle/control/hub.cpp [hunk 2]
```diff
@@ -1882,6 +1955,36 @@ void Hub::clearStatus()
...
+{
+    if (!Id::isModule(modId))
+        return Id::Invalid;
+    auto modName = m_stateTracker.getModuleName(modId);
+    for (size_t i = modId - 1; i >= message::Id::ModuleBase; --i) {
+        auto modit = m_stateTracker.runningMap.find(i);
+        if (modit != m_stateTracker.runningMap.end()) {
+            auto &av = m_stateTracker.getStaticModuleInfo(modit->first);
+        if (av.isCompound()) {
+                if (av.submodules().size() >= modId - i && av.submodules()[modId - i - 1].name == modName)
```
Comment (aumuell) [inline lib/vistle/control/hub.cpp:1968]: what if a compound contains two or more instance of the same module?

File: lib/vistle/core/messages.h [hunk 3]
```diff
@@ -435,10 +435,11 @@ class V_COREEXPORT AddObjectCompleted: public MessageBase<AddObjectCompleted, AD
     port_name_t m_orgDestPort;
 };
 
-//! connect an output port to an input port of another module
-class V_COREEXPORT Connect: public MessageBase<Connect, CONNECT> {
+//! Base class for connect and disconnect
+
```
Comment (aumuell) [inline lib/vistle/core/messages.h:439]: probably won't work portably due to unclear memory layout with multiple inheritance

File: lib/vistle/core/statetracker.cpp [hunk 4]
```diff
@@ -1846,6 +1833,17 @@ void StateTracker::updateStatus()
...
 
+const AvailableModule &StateTracker::getStaticModuleInfo(int modId)
+{
+    auto hubId = getHub(modId);
+    if (hubId != Id::Invalid) {
+        auto it = availableModules().find({hubId, getModuleName(modId)});
+        if (it != availableModules().end())
+            return it->second;
+    }
+    throw vistle::exception{"getStaticModuleInfo failed mor module with id " + std::to_string(modId)};
```
Comment (aumuell) [inline lib/vistle/core/statetracker.cpp:1844]: ...failed for module...

File: lib/vistle/control/vistle.py [hunk 5]
```diff
@@ -296,17 +296,23 @@ def load(filename = None):
...
 def moduleCompoundAddModule(compoundId, moduleName, x, y):
    return _vistle.moduleCompoundAddModule(compoundId, moduleName, x, y)
 
 def moduleCompoundConnect(compoundId, fromId, fromName, toId, toName): #if fromId or toId is equal to compound id the port is exposed with the given name
    _vistle.moduleCompoundConnect(compoundId, fromId, fromName, toId, toName)
 
-def moduleCompoundCreate(compoundId):
-   _vistle.moduleCompoundCreate(compoundId)
+def moduleCompoundExpose(compoundId, compoundPortName, modId, modPortName, portType):
+   _vistle.moduleCompoundExpose(compoundId, compoundPortName, modId, modPortName, portType)
```
Comment (aumuell) [inline lib/vistle/control/vistle.py:312]: how about having three of them?
- moduleCompoundExposeInput
- moduleCompoundExposeOutput
- moduleCompoundExposeParameter

File: lib/vistle/userinterface/pythonmodule.cpp [hunk 6]
```diff
@@ -858,10 +858,21 @@ static int moduleCompoundAlloc(const std::string &compoundName)
     }
 }
 
+
 static void compoundNotFoundError(int compoundId)
 {
     printError("module compound with id " + std::to_string(compoundId) + "hast not been found!");
-    printError("allocate it with moduleCompoundAlloc(\"compoundName\"");
+    printError("allocate it with moduleCompoundCreate(\"compoundName\"");
```
Comment (aumuell) [inline lib/vistle/userinterface/pythonmodule.cpp:865]: ...has not been found!

File: lib/vistle/userinterface/pythonmodule.cpp [hunk 7]
```diff
@@ -901,6 +912,26 @@ static void moduleCompoundConnect(int compoundId, int fromId, const std::string
...
+                                 const std::string &modPortName, vistle::Port::Type portType)
+{
+    auto comp = findCompound(compoundId);
+    if (comp == compounds.end()) {
+        compoundNotFoundError(compoundId);
+        return;
+    }
+    auto inputs = getInputPorts(modId);
+    auto outputs = getOutputPorts(modId);
+    std::cerr << "inputs " << inputs.size() << " outputs " << outputs.size() << std::endl;
```
Comment (aumuell) [inline lib/vistle/userinterface/pythonmodule.cpp:925]: remove unspecific debug message

File: lib/vistle/userinterface/pythonmodule.cpp [hunk 8]
```diff
@@ -956,26 +987,31 @@ void moduleCompoundToFile(const ModuleCompound &comp)
...
     }
     file << std::endl;
     for (const auto &conn: comp.connections()) {
-        file << "moduleCompoundConnect(CompoundId, " << conn.fromId << ",\"" << conn.fromPort << "\", " << conn.toId
+        std::string fromIdName = conn.fromId >= 0
+                                     ? "um" + comp.submodules()[conn.fromId].name + std::to_string(conn.fromId)
+                                     : std::to_string(conn.fromId);
+        std::string toIdName = conn.toId >= 0 ? "um" + comp.submodules()[conn.toId].name + std::to_string(conn.toId)
+                                              : std::to_string(conn.toId);
+        file << "moduleCompoundConnect(CompoundId, " << fromIdName << ",\"" << conn.fromPort << "\", " << toIdName
```
Comment (aumuell) [inline lib/vistle/userinterface/pythonmodule.cpp:1003]: use moduleCompoundExpose... instead

**Review by aumuell — CHANGES_REQUESTED** [review]
please look at the inline comments below

File: lib/vistle/control/hub.cpp (hunk 2 above)
Comment (hpcdgrie) [reply lib/vistle/control/hub.cpp:1968]: av.submodules().size() >= modId - sould always be a sufficent check. av.submodules()[modId - i - 1].name == modName ist just to be extra save.

Comment (aumuell) [reply lib/vistle/control/hub.cpp:1968]: make the last check an assert() then?

**Pull request closed without merging**